subprojects/docs/src/docs/userguide/upgrade_version_4.adoc

```diff
@@ -0,0 +1,493 @@
+ * `getTargetPlatform()` was changed to return a `Property`.
+ * `setTargetPlatform(NativePlatform)` was removed. Use `Property.set()` instead.
+ * `getToolChain()` was changed to return a `Property`.
+ * `setToolChain(NativeToolChain)` was removed. Use `Property.set()` instead.
+ * `getStaticLibArgs()` was changed to return a `ListProperty`.
+ * `setStaticLibArgs(List)` was removed. Use `ListProperty.set()` instead.
+
+link:{groovyDslPath}/org.gradle.nativeplatform.tasks.InstallExecutable.html[InstallExecutable]::
+ * `getSourceFile()` was replaced by `getExecutableFile()`.
+ * `getPlatform()` was replaced by `getTargetPlatform()`.
+ * `setTargetPlatform(NativePlatform)` was removed. Use `Property.set()` instead.
+ * `getToolChain()` was changed to return a `Property`.
+ * `setToolChain(NativeToolChain)` was removed. Use `Property.set()` instead.
+
+The following have also seen similar changes:
+
+ * link:{groovyDslPath}/org.gradle.language.assembler.tasks.Assemble.html[Assemble]
+ * link:{groovyDslPath}/org.gradle.language.rc.tasks.WindowsResourceCompile.html[WindowsResourceCompile]
+ * link:{javadocPath}/org/gradle/nativeplatform/tasks/StripSymbols.html[StripSymbols]
+ * link:{javadocPath}/org/gradle/nativeplatform/tasks/ExtractSymbols.html[ExtractSymbols]
+ * link:{javadocPath}/org/gradle/language/swift/tasks/SwiftCompile.html[SwiftCompile]
+ * link:{javadocPath}/org/gradle/nativeplatform/tasks/LinkMachOBundle.html[LinkMachOBundle]
+
+[[rel4.5:bom_import]]
+=== [4.5] BOM import
+
+Gradle now provides support for importing bill of materials (BOM) files, which are effectively POM files that use `<dependencyManagement>` sections to control the versions of direct and transitive dependencies. All you need to do is declare the POM as just another dependency.
+
+The following example picks the versions of the `gson` and `dom4j` dependencies from the declared Spring Boot BOM:
+
+----
+dependencies {
+    // import a BOM
+    implementation 'org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE'
+
+    // define dependencies without versions
+    implementation 'com.google.code.gson:gson'
+    implementation 'dom4j:dom4j'
+}
+----
+
+
+[[rel4.5:pom_optional_dependencies]]
+=== [4.5] Support for optional dependencies when consuming POMs
+
+Gradle now creates a <<managing_transitive_dependencies.adoc#sec:dependency_constraints,dependency constraint>> for each dependency declaration in a POM file with an `<optional>true</optional>` element. This results in the expected behavior:
+
+ * The dependency module is ignored if it is only ever declared as optional.
+ * If the dependency module is also declared elsewhere as not optional, then the constraint derived from the optional dependency declaration is considered when picking the version.
+
+In other words, if an optional dependency has a declared version higher than another, non-optional one, the optional dependency's version is used. However, Gradle does not resolve optional dependencies.
```

Comment:
> However, Gradle does not resolve optional dependencies.

I think that's the wrong terminology. It does resolve them, that's why version selection now works. What you're trying to say is probably: It doesn't make them part of the graph unless someone else has an actual dependency on them. But that's already mentioned in the paragraph above.

Reply:
Yes, but I'm trying to clarify the previous paragraph into terminology I find easier to understand. I'll have another go.